lib/MySQL_Thread.cpp

```diff
@@ -1464,3 +1872,30 @@ void MySQL_Threads_Handler::signal_all_threads() {
 	}
 }
 
+bool MySQL_Threads_Handler::kill_session(uint32_t _thread_session_id) {
+	bool ret=false;
+	unsigned int i;
+	signal_all_threads(1);
+	for (i=0;i<num_threads;i++) {
```

Comment:
More details here.
In MySQL_Thread::run() the main loop is:
```
while (shutdown==0) {
...
spin_wrunlock(&thread_mutex);
pool()
spin_wrlock(&thread_mutex);
...
}
```

What thread_mutex protects is *not* a shared memory, but it is protecting *all* the data structures of the worker thread .
If another thread gets a lock on that thread_mutex , it can access all the thread's data structures knowing it is safe to do so. That is used, for example, by processlist.
Right now it is also used by KILL , but this can be improved in future.

In other words, the mutex doesn't protect a shared memory, but it is a mechanism to completely stop the thread